### ISSUE TITLE
Sitemap not loading stylesheet on sub paths

### ISSUE THREAD

**Issue by brandsis** [tobimori/kirby-seo#88]
Hi, Tobias. I've installed v1.0.0 but I'm still not getting a sitemap to generate. I've added the blueprint tabs and meta snippets, and I've set the Robots Directives for each page in the Panel. `/robots.txt` is working fine but there is nothing showing at `/sitemap.xml` – it's a blank page with only `xml-stylesheet` in the Dev Tools Elements.

What am I missing?

This is what's in my `config.php` file...

```
	'tobimori.seo' => [
		'canonicalBase' => 'http://localhost/projects/brandsistency',
		'lang' => 'en_GB',
		'robots' => [
			'active' => true,
			'content' => [
				'*' => [
					'Allow' => ['/'],
					'Disallow' => ['/kirby', '/panel', '/content']
				]
			],
			'sitemap' => 'http://localhost/projects/brandsistency/sitemap.xml'
		],
		'sitemap' => [
			'active' => true,
			'excludeTemplates' => ['tertiary', 'maintenance']
		]
	],
```

Thanks.

**Comment by tobimori** [comment]
What's the output of `viewsource:http://localhost/projects/brandsistency/sitemap.xml`?

**Comment by tobimori** [comment]
Besides that, it shouldn't be necessary specify `robots.active`, `sitemap.active` & `robots.sitemap`.

**Comment by brandsis** [comment]
> What's the output of `viewsource:http://localhost/projects/brandsistency/sitemap.xml`?

There is no output, just a blank page. If I right-click, there is no context menu. In Dev Tools, the Elements section contains only `xml-stylesheet` and the Console has the error `GET http://localhost/sitemap.xsl net::ERR_ABORTED 404 (Not Found)`.

From that, I tried `/sitemap.xsl` and there is a styled page there but the sitemap is empty...

![image](https://github.com/tobimori/kirby-seo/assets/153773853/d5acf68a-2704-4f57-9e02-4199f0f808f6)

**Comment by tobimori** [comment]
It seems like the issue is that Chrome is trying to get the sitemap stylesheet from `http://localhost/sitemap.xsl` but your project is in a subfolder, so it's located at `http://localhost/projects/brandsistency/sitemap.xsl` instead.

**Comment by brandsis** [comment]
No change.

I updated line 45 in the file `http://localhost/projects/brandsistency/site/plugins/seo/classes/Sitemap/SitemapIndex.php`, cleared cache in the Panel and the browser, did a hard reload and the result was the same – still no `/sitemap.xml` output and the same error message in the Console. I also tried heard coding the `href` URL in that line and again no change.

**Comment by tobimori** [comment]
hm, that's weird

some other things to consider:
- does it happen with a new plainkit locally?
- does it work in a different browser? 
- does it happen without the subpath? 
- does it happen on a production server with a subpath?

**Comment by brandsis** [comment]
Tobias, it will take me a little while to answer all those questions. I'll get back to you soon.

**Comment by brandsis** [comment]
Hi, Tobias.

I haven't tried a fresh plainkit install but I can answer the other questions.

* **does it work in a different browser?**

I get the same result in most other browsers. In Firefox Developer Edition, there is a different result, albeit not a successful one. If I go to `localhost/projects/brandsistency/sitemap.xsl`, it downloads the file rather than displaying on screen. And if I go to `localhost/projects/brandsistency/sitemap.xml`, instead of a completely blank page, I see this...

![image](https://github.com/tobimori/kirby-seo/assets/153773853/f91feff6-489c-4a49-a99d-8898d9968f05)

It still seems to be looking in the root for the `.xsl` file, despite that update to line 45 in `SitemapIndex.php`.

* **does it happen on a production server with a subpath?**

Yes. I promoted the changes to my test site (brandsistency.dev/brandsistency), which is public-facing site but also in a subfolder, and the results were the same. [brandsistency.dev/brandsistency/sitemap.xml](https://brandsistency.dev/brandsistency/sitemap.xml) is blank, or in Firefox Dev Ed gives a similar screen looking for the `.xsl` file in the location `https://brandsistency.dev/sitemap.xsl` – i.e. in the root again not in the subfolder.

* **does it happen without the subpath?**

No. I also promoted the updates to my live site (brandsistency.com), which is both public-facing and located at the domain root. Everything is working fine there. [brandsistency.com/sitemap.xml](https://brandsistency.com/sitemap.xml) looks as expected...

![image](https://github.com/tobimori/kirby-seo/assets/153773853/770b0855-28a2-4042-ab91-dd4ba34483fa)

So, as you thought, it's only an issue with subfolder sites.

Not a massive deal as it's working in production, but it would be good to have it working in subfolder sites. That's so clients can be reassured that the sitemap is working when looking at a test site before going to production. 🙂

**Comment by tobimori** [comment]
Thank you very much. It seems like it's indeed an issue with the stylesheet, so it should not prevent indexing by crawlers even on a subpath in production.

If you open `view-source:https://brandsistency.dev/brandsistency/sitemap.xml` in your Chrome, you'll see the source code as expected.

**Comment by tobimori** [comment]
What's confusing for me is that you said even with changing the `<?xml-stylesheet type="text/xsl" href="/sitemap.xsl"?>` manually to the absolute url, e.g.`https://brandsistency.dev/brandsistency/sitemap.xsl` it's not working for you?

**Comment by brandsis** [comment]
> What's confusing for me is that you said even with changing the `<?xml-stylesheet type="text/xsl" href="/sitemap.xsl"?>` manually to the absolute url, e.g.`https://brandsistency.dev/brandsistency/sitemap.xsl` it's not working for you?

I didn't actually try hard-coding the URL in the test (.dev) environment – only in localhost. I'll try that later today.

**Comment by brandsis** [comment]
Tobias, I have tried hard-coding the stylesheet location in `SitemapIndex.php`. Line 45 now reads...

`$doc->appendChild($doc->createProcessingInstruction('xml-stylesheet', 'type="text/xsl" href="https://brandsistency.dev/brandsistency/sitemap.xsl"'));`

I've cleared cache in the Panel and the browser and there is no change...

* `https://brandsistency.dev/brandsistency/sitemap.xml` is blank with only "xml-stylesheet" showing in Dev Tools
* `view-source:https://brandsistency.dev/brandsistency/sitemap.xml` does still contain the sitemap content
* `https://brandsistency.dev/brandsistency/sitemap.xml` is the styled but empty view

**Comment by tobimori** [comment]
![CleanShot 2024-03-26 at 10 01 10@2x](https://github.com/tobimori/kirby-seo/assets/29142128/7f942114-ddeb-4af0-affe-05aef35aae0a)

Was the change made on the URL you linked or did you remove it again? It shows the /sitemap.xsl without the full path here.

**Comment by brandsis** [comment]
No, I didn't remove it again. This is Line 45 right now, viewed through cPanel...

![image](https://github.com/tobimori/kirby-seo/assets/153773853/041c2ff9-2e6f-4a23-8229-9919e8510897)

**Comment by tobimori** [comment]
Have you cleared OP Cache?

**Comment by brandsis** [comment]
No, I haven't. The only way I'm aware of is to create a file called `flush_cache.php` with the code...

```
<?php
opcache_reset();
?>
```
...to upload that to the site root and then call it via the browser. I did that but `https://brandsistency.dev/brandsistency/flush_cache.php` returns an error and when I visit the sitemap it is still blank.

Any other suggestions on how I can clear OPcache?

**Comment by tobimori** [comment]
Except for Kirby Pages cache and OPcache - no idea where else it should be cached.

**Comment by brandsis** [comment]
Tobias, I found out that the "opcache" extension wasn't enabled in my hosting account's PHP config. I have now enabled it but the `flush_cache.php` script only seems to work from my account root. I called the file from that location and did not get the error message, so I'm assuming it works. Not sure how I check, though.

Calling that file/script from any of my 'add-on' domains, such as brandsistency.dev, does still throw the error. But I am told by my host (Namecheap) that flushing from the account root should work for all domains in my account.

Anyway, I have cleared Panel and browser cache again and tried accessing `https://brandsistency.dev/brandsistency/sitemap.xml` but it is still blank. The `href` in the source still does not show the full path.

**Comment by nilshoerrmann** [comment]
> What's confusing for me is that you said even with changing the `<?xml-stylesheet type="text/xsl" href="/sitemap.xsl"?>` manually to the absolute url, e.g.`https://brandsistency.dev/brandsistency/sitemap.xsl` it's not working for you?

Hardcoding the path – not the full URL – works for me.

**Comment by Jobbies** [comment]
I am also having the same problem. Could this be from the routes.php? 

```
	[
		'pattern' => 'sitemap',
		'action' => function () {
			if (!option('tobimori.seo.sitemap.redirect', true) || !option('tobimori.seo.sitemap.active', true)) {
				$this->next();
			}

			go('/sitemap.xml');
		}
	],
```

Should the Route be dynamic to the site? 

```
go(site()->url() . '/sitemap.xml');
```

**Comment by MelleNi** [comment]
So does sitemap not work for anyone? For me it doesn't seem to generate anything, and going to `https://mydomain.com/sitemap.xsl` just downloads a default file.

**Comment by tobimori** [comment]
The sitemap stylesheet does not matter for the actual visibility and usefulness of a sitemap for search engines, as long as you're able to go to `view-source:https://mydomain.com/sitemap.xml` and it is shown. The XSL file itself would show "buggy" in browsers or download (depending on your server configuration, so if the content-type file type is `text/xml` or `application/xslt+xml`)

**Comment by MelleNi** [comment]
Ah right. It's not generating a sitemap.xml to begin with. Just updated to your latest version and still the same, not sure how to trigger it?

**Comment by tobimori** [comment]
It should work, I assume something else must be wrong with your setup.

**Comment by MelleNi** [comment]
Any method you recommend to test what is wrong with my setup?

**Comment by tobimori** [comment]
Unfortunately, no idea.

**Comment by MelleNi** [comment]
I got it to work, but this `<urlset xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">` at the top, is incorrect according to google, and they won't parse it
